pkg/resource/deploy/source_eval.go

```diff
@@ -1525,6 +1525,27 @@ func transformAliasForNodeJSCompat(alias *pulumirpc.Alias) *pulumirpc.Alias {
 	return alias
 }
 
+func (rm *resmon) resolveProvider(
+	provider string, providers map[string]string, parent resource.URN, pkg tokens.Package,
+) string {
+	if provider != "" {
+		return provider
+	}
+	if prov, ok := providers[string(pkg)]; ok {
+		return prov
+	}
+	if parent != "" {
+		rm.componentProvidersLock.Lock()
+		defer rm.componentProvidersLock.Unlock()
+		if parentsProvider, ok := rm.componentProviders[parent]; ok {
+			if prov, ok := parentsProvider[string(pkg)]; ok {
+				return prov
+			}
```

Comment:
Guessing this doesn't need to be recursive since we'll have already processed the parent's parent in a previous call to `RegisterResource` and updated the `componentProviders`, right?

Reply:
Yes, that's my understanding from the code.